File: native/tests/VolumeCurveTests.cpp
#include "doctest.h"
#include "winmix/audio/VolumeCurve.h"

#include <limits>
#include <stdexcept>

using winmix::audio::VolumeCurve;

TEST_CASE("ZeroPositionIsTrueSilence")
{
    // Not just the dB floor: a slider pulled to the bottom must be inaudible.
    CHECK(VolumeCurve::ToScalar(0.0) == 0.0f);
}

TEST_CASE("FullPositionIsUnityGain")
{
    CHECK(VolumeCurve::ToScalar(1.0) == doctest::Approx(1.0f).epsilon(0.00001));
}

TEST_CASE("MidPositionSitsWellBelowHalfAmplitude")
{
    // The whole point of the taper: half travel is roughly -30 dB, not 0.5
    // amplitude. If this ever reads ~0.5 the curve has been bypassed.
    const float scalar = VolumeCurve::ToScalar(0.5);

    CHECK(scalar >= 0.01f);
    CHECK(scalar <= 0.1f);
}

TEST_CASE("PositionSurvivesRoundTrip")
{
    for (const double position : {0.05, 0.25, 0.5, 0.75, 1.0})
    {
        const double restored = VolumeCurve::ToPosition(VolumeCurve::ToScalar(position));
        CHECK(restored == doctest::Approx(position).epsilon(0.0001));
    }
}

TEST_CASE("CurveIsMonotonic")
{
    float previous = -1.0f;

    for (double position = 0.0; position <= 1.0; position += 0.01)
    {
        const float scalar = VolumeCurve::ToScalar(position);
        CHECK_MESSAGE(scalar >= previous, "Curve dipped at position " << position);
        previous = scalar;
    }
}

TEST_CASE("PositionsOutsideUnitRangeAreClamped")
{
    CHECK(VolumeCurve::ToScalar(-0.5) == doctest::Approx(0.0f));
    CHECK(VolumeCurve::ToScalar(1.5) == doctest::Approx(1.0f));
}

TEST_CASE("ScalarsBelowTheFloorCollapseToZeroPosition")
{
    const double belowFloor = VolumeCurve::DbToScalar(VolumeCurve::kDefaultMinDb - 10.0);
    CHECK(VolumeCurve::ToPosition(belowFloor) == 0.0);
}

TEST_CASE("ASteeperFloorPushesTheSamePositionQuieter")
{
    const float gentle = VolumeCurve::ToScalar(0.5, -30.0);
    const float steep = VolumeCurve::ToScalar(0.5, -90.0);
    CHECK(steep < gentle);
}

TEST_CASE("ANonNegativeFloorIsRejected")
{
    for (const double minDb : {0.0, 6.0, std::numeric_limits<double>::quiet_NaN()})
    {
        CHECK_THROWS_AS(VolumeCurve::ToScalar(0.5, minDb), std::invalid_argument);
        CHECK_THROWS_AS(VolumeCurve::ToPosition(0.5, minDb), std::invalid_argument);
    }
}

TEST_CASE("SilenceHasNoDecibelValue")
{
    CHECK(VolumeCurve::ScalarToDb(0.0) == -std::numeric_limits<double>::infinity());
}

TEST_CASE("DecibelsConvertToTheExpectedAmplitude")
{
    CHECK(VolumeCurve::DbToScalar(0.0) == doctest::Approx(1.0).epsilon(0.001));
    CHECK(VolumeCurve::DbToScalar(-6.0) == doctest::Approx(0.501).epsilon(0.001));
    CHECK(VolumeCurve::DbToScalar(-20.0) == doctest::Approx(0.1).epsilon(0.001));
}

File: tests/WinMix.Audio.Tests/VolumeCurveTests.cs
using WinMix.Audio;
using Xunit;

namespace WinMix.Audio.Tests;

public class VolumeCurveTests
{
    [Fact]
    public void ZeroPositionIsTrueSilence()
    {
        // Not just the dB floor: a slider pulled to the bottom must be inaudible.
        Assert.Equal(0f, VolumeCurve.ToScalar(0.0));
    }

    [Fact]
    public void FullPositionIsUnityGain()
    {
        Assert.Equal(1f, VolumeCurve.ToScalar(1.0), precision: 5);
    }

    [Fact]
    public void MidPositionSitsWellBelowHalfAmplitude()
    {
        // The whole point of the taper: half travel is roughly -30 dB, not 0.5
        // amplitude. If this ever reads ~0.5 the curve has been bypassed.
        var scalar = VolumeCurve.ToScalar(0.5);

        Assert.InRange(scalar, 0.01f, 0.1f);
    }

    [Theory]
    [InlineData(0.05)]
    [InlineData(0.25)]
    [InlineData(0.5)]
    [InlineData(0.75)]
    [InlineData(1.0)]
    public void PositionSurvivesRoundTrip(double position)
    {
        var restored = VolumeCurve.ToPosition(VolumeCurve.ToScalar(position));

        Assert.Equal(position, restored, precision: 4);
    }

    [Fact]
    public void CurveIsMonotonic()
    {
        var previous = -1f;

        for (var position = 0.0; position <= 1.0; position += 0.01)
        {
            var scalar = VolumeCurve.ToScalar(position);
            Assert.True(scalar >= previous, $"Curve dipped at position {position}.");
            previous = scalar;
        }
    }

    [Theory]
    [InlineData(-0.5, 0f)]
    [InlineData(1.5, 1f)]
    public void PositionsOutsideUnitRangeAreClamped(double position, float expected)
    {
        Assert.Equal(expected, VolumeCurve.ToScalar(position), precision: 5);
    }

    [Fact]
    public void ScalarsBelowTheFloorCollapseToZeroPosition()
    {
        var belowFloor = VolumeCurve.DbToScalar(VolumeCurve.DefaultMinDb - 10.0);

        Assert.Equal(0.0, VolumeCurve.ToPosition(belowFloor));
    }

    [Fact]
    public void ASteeperFloorPushesTheSamePositionQuieter()
    {
        var gentle = VolumeCurve.ToScalar(0.5, minDb: -30.0);
        var steep = VolumeCurve.ToScalar(0.5, minDb: -90.0);

        Assert.True(steep < gentle);
    }

    [Theory]
    [InlineData(0.0)]
    [InlineData(6.0)]
    [InlineData(double.NaN)]
    public void ANonNegativeFloorIsRejected(double minDb)
    {
        Assert.Throws<ArgumentOutOfRangeException>(() => VolumeCurve.ToScalar(0.5, minDb));
        Assert.Throws<ArgumentOutOfRangeException>(() => VolumeCurve.ToPosition(0.5, minDb));
    }

    [Fact]
    public void SilenceHasNoDecibelValue()
    {
        Assert.Equal(double.NegativeInfinity, VolumeCurve.ScalarToDb(0.0));
    }

    [Theory]
    [InlineData(0.0, 1.0)]
    [InlineData(-6.0, 0.501)]
    [InlineData(-20.0, 0.1)]
    public void DecibelsConvertToTheExpectedAmplitude(double db, double expected)
    {
        Assert.Equal(expected, VolumeCurve.DbToScalar(db), precision: 3);
    }
}
